src/coreComponents/dataRepository/Wrapper.hpp

```diff
@@ -912,6 +800,137 @@ class Wrapper final : public WrapperBase
 #endif
 
 private:
+
+  /**
+   * @brief Concrete implementation of the packing method.
+   * @tparam DO_PACKING A template parameter to discriminate between actually packing or only computing the packing size.
+   * @param[in,out] buffer The buffer that will receive the packed data.
+   * @param[in] withMetadata Whether to pack string metadata with the underlying data.
+   * @param[in] onDevice Whether to use device-based packing functions
+   *                     (buffer must be either pinned or a device pointer)
+   * @param[out] events A collection of events to poll for completion of async
+   *                    packing kernels ( device packing is incomplete until all
+   *                    events are finalized )
+   * @return The packed size.
+   */
+  template< bool DO_PACKING >
+  localIndex packTemplate( buffer_unit_type * & buffer,
```

Comment:
The previous pattern for `pack` (and `unpack`, it's the same) was having pure virtual `WrapperBase::pack`/`WrapperBase::packSize` member functions, delegating the implementations to `Wrapper< T >`.

To keep the symmetry with other code and remove some duplicated code, I've added a `WrapperBase::pack< DO_PACKING >` as a `public` member function. To keep the delegation pattern, I have pure virtual `WrapperBase::packSizePrivate`/`WrapperBase::packPrivate`.

Then `Wrapper` has to implement `WrapperBase::packSizePrivate`/`WrapperBase::packPrivate`. To remove code duplication I went on implementing `private` `WrapperBase::packTemplate< DO_PACKING >`, because of name conflict, which could admittedly be called `WrapperBase::packImpl< DO_PACKING >` (or something else)?

Reply:
I've pushed a version with `packImpl` with is a little better I think. Is it OK?

Reply:
@TotoGaz I think that is a little more accurate of a name, don't you? Perhaps we should adopt this naming convention for this pattern in general.

Reply:
I do not think that this exact pattern (double indirection) actually happens too often (maybe it's the only one?).
I think the idea was to be able to call the `pack/unpack` from the `base` class without using some `CRTP`.

Sometime we can see some `packPrivate` (with a single indirection), which could be renamed into `packImpl`.